overlord/servicestate/service_control.go

```diff
@@ -87,15 +87,19 @@ func (m *ServiceManager) doServiceControl(t *state.Task, _ *tomb.Tomb) error {
 		}
 	}
 
-	meter := snapstate.NewTaskProgressAdapterLocked(t)
+	meter := snapstate.NewTaskProgressAdapterUnlocked(t)
 
+	// Note - state must be unlocked when calling wrappers below.
 	switch sc.Action {
 	case "stop":
 		disable := sc.ActionModifier == "disable"
 		flags := &wrappers.StopServicesFlags{
 			Disable: disable,
 		}
-		if err := wrappers.StopServices(services, flags, snap.StopReasonOther, meter, perfTimings); err != nil {
+		st.Unlock()
+		err := wrappers.StopServices(services, flags, snap.StopReasonOther, meter, perfTimings)
+		st.Lock()
+		if err != nil {
```

Comment:
given that we might change and write snapst back,  the safest is probably to read it again after we have reacquired the lock

Reply:
Updated.